File: app/src/main/kotlin/com/openclaw/ghostcrab/domain/repository/ScopeProbe.kt
package com.openclaw.ghostcrab.domain.repository

/**
 * Runs `auth.whoami` over WebSocket to learn what scopes the current token carries.
 *
 * Treats "unknown" (old gateways without the method) as a distinct state so callers
 * can fall back to the Phase-4 copy-paste install instructions instead of showing a
 * misleading "scope missing" error.
 */
interface ScopeProbe {
    /** @return result; never throws. */
    suspend fun probe(): ScopeProbeResult
}

sealed interface ScopeProbeResult {
    data class Known(val scopes: Set<String>) : ScopeProbeResult {
        fun has(scope: String): Boolean = scope in scopes
    }
    /** Gateway did not implement `auth.whoami` — treat feature as "unknown capability". */
    data object UnknownOldGateway : ScopeProbeResult
    /** Probe failed (network etc.). Caller decides UX. */
    data class Failed(val cause: String) : ScopeProbeResult
}

File: ios/GhostCrab/Domain/Repository/ScopeProbe.swift
import Foundation

/// Runs `auth.whoami` over WebSocket to learn what scopes the current token carries.
///
/// Treats "unknown" (old gateways without the method) as a distinct state so callers
/// can fall back to the Phase-4 copy-paste install instructions instead of showing a
/// misleading "scope missing" error.
public protocol ScopeProbe: Sendable {
    /// - Returns: A `ScopeProbeResult` describing the outcome. Never throws.
    func probe() async -> ScopeProbeResult
}

/// Result of a scope probe.
///
/// Direct port of the Kotlin `sealed interface ScopeProbeResult`.
public enum ScopeProbeResult: Sendable {

    /// The gateway responded with a concrete scope set.
    case known(scopes: Set<String>)

    /// Gateway did not implement `auth.whoami` — treat feature as "unknown capability".
    case unknownOldGateway

    /// Probe failed (network etc.). Caller decides UX.
    case failed(cause: String)

    /// Convenience: whether a `.known` result contains the given scope. Returns `false` for
    /// `.unknownOldGateway` and `.failed`.
    public func has(_ scope: String) -> Bool {
        if case .known(let scopes) = self { return scopes.contains(scope) }
        return false
    }
}
